cpp/src/stream_compaction/stream_compaction.cu

```diff
@@ -108,4 +114,136 @@ gdf_column drop_nulls(gdf_column const &input) {
+                             InputIterator last,
+                             OutputIterator output,
+                             BinaryPredicate comp,
+                             const duplicate_keep_option keep)
+{
+  IndexType n = (last-first)-1;
+  if (keep == duplicate_keep_option::KEEP_FIRST) {
+      return thrust::copy_if(exec,
+              first,
+              last,
+              thrust::counting_iterator<IndexType>(0),
+              output,
+              [first, comp, n] __device__ (const IndexType i) {
+              return (i == 0 || !comp(first[i], first[i-1]));
+              });
+  } else if (keep == duplicate_keep_option::KEEP_LAST) {
+      return thrust::copy_if(exec,
+              first,
+              last,
+              thrust::counting_iterator<IndexType>(0),
+              output,
+              [first, comp, n] __device__ (const IndexType i) {
+              return (i == n || !comp(first[i], first[i+1]));
+              });
+  } else {
+      return thrust::copy_if(exec,
+              first,
+              last,
+              thrust::counting_iterator<IndexType>(0),
+              output,
+              [first, comp, n] __device__ (const IndexType i) {
+              return (i == 0 || !comp(first[i], first[i-1])) 
+                  && (i == n || !comp(first[i], first[i+1]));
+              });
+  }
+}
+
+auto 
+get_unique_ordered_indices(const cudf::table& key_columns,
+                           const duplicate_keep_option keep,
+                           cudaStream_t stream=0)
+{
+  gdf_size_type ncols = key_columns.num_columns();
+  gdf_size_type nrows = key_columns.num_rows();
+
+  // sort only indices
+  rmm::device_vector<gdf_size_type> sorted_indices(nrows);
+  gdf_context context;
+  gdf_column sorted_indices_col;
+  CUDF_TRY(gdf_column_view(&sorted_indices_col, (void*)(sorted_indices.data().get()),
+        nullptr, nrows, GDF_INT32));
+  CUDF_TRY(gdf_order_by(key_columns.begin(),
+        nullptr,
+        key_columns.num_columns(),
+        &sorted_indices_col,
+        &context));
+
+  // extract unique indices 
+  rmm::device_vector<gdf_index_type> unique_indices(nrows);
+  auto exec = rmm::exec_policy(stream)->on(stream);
+  auto device_input_table = device_table::create(key_columns, stream);
+  rmm::device_vector<gdf_size_type>::iterator result_end;
+
+  bool nullable = device_input_table.get()->has_nulls();
+  if(nullable) {
+    auto comp = row_equality_comparator<true>(*device_input_table, true);
+    result_end = unique_copy(exec,
+        sorted_indices.begin(),
+        sorted_indices.end(),
+        unique_indices.begin(),
+        comp,
+        keep);
+  } else {
+    auto comp = row_equality_comparator<false>(*device_input_table, true);
+    result_end = unique_copy(exec,
+        sorted_indices.begin(),
+        sorted_indices.end(),
+        unique_indices.begin(),
+        comp,
+        keep);
+  }
+ 
+  //not resizing vector to avoid copy
+  cudaStreamSynchronize(stream);
```

Comment:
Don't sync the stream. Once we expose streams, it will be up to the user to sync the stream

Reply:
removed sync stream.